SEED-platform/seed
# Feat/bsync parser

#### Any background context you want to provide?
The BuildingSync parser and general import flow needs to be refactored to handle multiple schema versions.
This is one part of of a couple steps to improve the import process of these files
#### What's this PR do?
- Replaces the BuildingSync parser and exporter with more generalized solutions, allowing us to handle different schema versions
- allows uploading and exporting of v2.0 buildingsync files
- adds new application scales to property measures from bsync schema v2 (common areas, tenant areas, entire building)

NOTE: this will BREAK the exporting of files already on BRICR. After merging this PR, you must run the transformations on the file in the repo [here](https://github.com/BuildingSync/transformations).

#### How should this be manually tested?
##### Setup
Pull this branch. A python package and migration was added so apply migrations and reinstall requirements.
##### Test
- upload ex_1_no_schemaLocation.xml (this branch at seed/building_sync/tests/data). This file is missing the expected BuildingSync schema URL so it should fail with an error about no schema location being found
- uploade ex_1_no_address.xml. The file is missing address info, so should make it past the mapping page, but fails once you try to map it.
- upload buildingsync_v2_0_bricr_workflow.xml
    - check property info looks correct
    - check scenarios are there, check meter data
    - export the BuildingSync file, there should be no significant differences between the original and the exported file
- upload buildingsync_ATT_export.xml; tests that we can upload files that were exported from Audit Template Tool.
    - verify property info looks correct
    - export BuildingSync file, there should be no significant differences between the original and the exported file
- upload a file with non-required fields (ex1.xml, it has lat and long), then _updating_ that property with another file which doesn't have those fields (ex1_no_lat_long.xml). Export the property and it should have the lat and long in the xml. Then run the exported file thorough a schema validator with schema 2.0 (this schema can be found at /seed/building_sync/schemas, and if on OSX you can run `xmllint <exported_file> --schema <schema file>`). This is testing that we can pull data from the property state and that we can build valid xml.
- upload a NON buildingsync property, then export it as a BuildingSync xml. Verify that any property-state data the source file included is also a part of the exported xml.
#### What are the relevant tickets?
#1935 
#1985
